### DISCUSSION TITLE
adding a tag from a file

### DISCUSSION THREAD

**Discussion by BrianHicks** [cue-lang/cue#1582] [Q&A]
Hey folks! My CD system stores information about what container image tag to deploy in files next to the CUE source, and expects them to be set. Currently we're using `@tag(tag)` in the source, and a build script like this to get our Kubernetes manifests:

```sh
cue export \
    --out text \
    --expression 'yaml.MarshalStream(manifest)' \
    --inject "tag=$(cat tag)"
```

This works, but I can't help but feel that there's a better way. I see mentions of loading CUE-parseable files, but haven't ever been able to actually make it work. If I have a tag as text, can I load it into my manifests without using the tags system? Would it have to be JSON? What is the actual mechanism here?

I hope this is obvious to someone else, but it isn't to me—I've searched through the documentation trying to figure this out and have come up empty a couple times now.

**Comment by dave08** [comment]
In our CI (Drone), we check out the repo with the cue manifests, `cd` into the app's folder, and then run `echo -e "package kube\n\n#DockerTag: \"$DRONE_COMMIT\"" > version.cue`. And in concat `#DockerTag` in the image text and then commit that again.

**Comment by mpvl** [comment]
Would you want to inject the file on the command line, or would you rather inject the file within the CUE file (akin to Go's file embedding mechanism)?

**Reply by BrianHicks** [reply]
I'm injecting it on the command line now, but I'd kind of rather not. I'm also not sure how Go's file embedding mechanism works to answer that question.

So hopefully this will be helpful? We're moving away from Dhall and toward CUE. Right now in a Dhall file I can do this:

```dhall
{ someRecordKey = ./image as Text }
```

and it'll just read the file called `./image` relative to the current source file. I'd like to do something similar in CUE—everything's on disk, but I'm having trouble figuring out how to get CUE to slurp the file in with a generic set of command-line flags! (e.g. not `--inject`, but something else like loading stuff from JSON files in the package directory.)

**Reply by mpvl** [reply]
So it seems that the Go-style embed approach would work. Basically, this would allow JSON etc to be converted to CUE in a pre-processing stage and injected in CUE at the given location. So it would not be a language feature per se, but rather a preprocessing step.

Note that it is important property for it to be a preprocessing step. Dynamic imports make things quite confusing fast.

**Reply by BrianHicks** [reply]
Is this the mechanism you're talking about? https://pkg.go.dev/embed

If so, I could see that working well! I'd want to be able to specify the schema that the embedded value had to follow, though.

**Reply by ansel1** [reply]
Could the embedded/inject value just be bytes?  Generating ConfigMaps in cue with data static contents (in various types of text formats, like env/toml/xml/etc) means I need to put that static content inside a .cue file, which is kind of a drag (my editor doesn't know to apply syntax highlight and can't format it, for example).  If cue had something like go's embed feature, I could define the contents of the config map in separate files, and get something very similar to kustomize ConfigMap generators.

**Reply by rusty-jules** [reply]
We desperately desire an `@embed` attribute, specifically for scripts that are placed into k8s configmaps or jobs to get syntax highlighting and avoid the spacing issues in multi-line blocks (save you if you don't use tabs by default). Using tags for this breaks our structure because tags must be passed, but we have standardized tags across our cue modules for automation and deployment whereas `@embed` is always specific to a given module's fields
